californium-core/src/main/java/org/eclipse/californium/core/network/stack/ExchangeCleanupLayer.java

```diff
@@ -72,9 +73,12 @@ public void sendRequest(final Exchange exchange, final Request request) {
 	 */
 	@Override
 	public void sendResponse(final Exchange exchange, final Response response) {
-
-		if (response.isConfirmable() && !response.isNotification()) {
-			response.addMessageObserver(new CleanupMessageObserver(exchange));
+		if (!response.isNotification()) {
+			Type type = response.getType();
+			if (type == null || type == Type.CON) {
+				// if type is set later, add the cleanup preventive
+				response.addMessageObserver(new CleanupMessageObserver(exchange));
```

Comment:
is there drawback to add this (and so close exchange) even on none CON response ?

Reply:
Beside of a small performance drawback [Exchange.complete - remote origin](https://github.com/eclipse/californium/blob/master/californium-core/src/main/java/org/eclipse/californium/core/network/Exchange.java#L830) I can't see any malfunction. The unit tests are still green. Or do I miss something?

Reply:
I think it's OK too.
Eventually, if we want a behavior closest than before maybe we can check if message is CON in CleanupMessageObserver before to complete but this means more code and not sure there is any concrete benefits.